R/SandwichLayerVariance.R

```diff
@@ -558,3 +559,459 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
+  cov1 <- tapply(1:nrow(XX[zobs==1,]), bid[zobs==1], function(s) cov(XX[zobs==1,][s,]))
+  cov01 <- lapply(1:length(cov0), function(s) .add_mat_diag(cov0[[s]], cov1[[s]]))
+  
+  cov01u <- lapply(1:length(cov0), 
+                   function(s) if (!is.na(cov01[[s]][1,1])) cov01[[s]]
+                   else .add_mat_sqdif(XX, zobs, bid, s))
+  cov01l <- lapply(1:length(cov0), 
+                   function(s) if (!is.na(cov01[[s]][1,1])) -cov01[[s]]
+                   else -.add_mat_sqdif(XX, zobs, bid, s, upper=FALSE))
+  V01u <- Reduce('+', cov01u)
+  V01l <- Reduce('+', cov01l)
+  return(rbind(V01u, V01l))
+}
+
+#' @title (Internal) Helper function for design-based meat matrix calculation
+#' @keywords internal
+.add_mat_sqdif <- function(X, zobs, bid, b, upper = TRUE){
+  A <- X[zobs==0 & bid==b, ]
+  if (upper) B <- X[zobs==1 & bid==b, ] else B <- -X[zobs==1 & bid==b, ]
+  cov01u <- matrix(0, nrow=ncol(A), ncol=ncol(A))
+  for (j in 1:ncol(B)){
+    for (h in 1:nrow(B)){
+      cov01u[,j] <- cov01u[,j] + colSums((A + B[h,j])^2)
+    }
+  }
+  cov01u <- cov01u / nrow(A) / nrow(B)
+  return(cov01u)
+}
+
+#' @title (Internal) Calculate grave{phi}
+#' @keywords internal
+#' @param x a fitted \code{teeMod} model
+.get_phi_tilde <- function(x, ...){
+  ws <- x$weights
+  n <- length(ws) # number of units in Q
+  design_obj <- x@Design
+  df <- x$call$data
+  
+  # treatment assignments
+  name_trt <- colnames(design_obj@structure)[design_obj@column_index == "t"]
+  assignment <- df[[name_trt]]
+  k <- length(unique(assignment))
+  z_ind <- sapply(c(0,1), function(val) as.integer(assignment == val))
+  
+  # stratum ids 
+  name_blk <- colnames(design_obj@structure)[design_obj@column_index == "b"]
+  df <- .merge_block_id_cols(df, name_blk)
+  name_blk <- name_blk[1]
+  stratum <- df[[name_blk]]
+  s <- length(unique(stratum)) # number of blocks
+  b_ind <- sapply(unique(stratum), function(val) as.integer(stratum == val)) 
+  
+  # nuisance parameters p
+  wb <- matrix(replicate(s, ws), ncol = s) * b_ind
+  p <- t(z_ind) %*% wb / (matrix(1, nrow = k, ncol = n) %*% wb)
+  p1 <- p[2, ]
+  
+  # calculate phi tilde
+  phitilde <- matrix(nrow = n, ncol = s)
+  for (j in 1:s){
+    phitilde[,j] <- ws * (z_ind[,2] - p1[j]) * b_ind[,j]
+  }
+  return(phitilde)
+}
+
+#' @title (Internal) Product of \eqn{A_{pp}^{-1} A_{\tau p}^T}
+#' @keywords internal
+#' @param x a fitted \code{teeMod} model
+#' @return An \eqn{s\times k} matrix \eqn{A_{pp}^{-1} A_{\tau p}^T}
+.get_appinv_atp <- function(x, ...){
+  ws <- x$weights
+  # estimated treatment effect tau1 <- x$coefficients
+  n <- length(ws) # number of units in Q(?)
+  design_obj <- x@Design
+  df <- x$call$data
+  
+  name_trt <- colnames(design_obj@structure)[design_obj@column_index == "t"]
+  assignment <- df[[name_trt]]
+  k <- length(unique(assignment))
+  
+  # stratum ids 
+  name_blk <- colnames(design_obj@structure)[design_obj@column_index == "b"]
```

Comment:
Same `var_names()` notes

Reply:
Updated in commit https://github.com/benbhansen-stats/propertee/commit/6eb8927664674088f61b3323683e7d592a2de431